samza-core/src/main/java/org/apache/samza/processor/StreamProcessor.java

```diff
@@ -70,84 +201,83 @@
+  @VisibleForTesting
+  StreamProcessor(Config config, Map<String, MetricsReporter> customMetricsReporters, Object taskFactory,
+                  StreamProcessorLifecycleListener processorListener, JobCoordinator jobCoordinator) {
+    this.taskFactory = taskFactory;
+    this.config = config;
+    this.taskShutdownMs = new TaskConfigJava(config).getShutdownMs();
+    this.customMetricsReporter = customMetricsReporters;
+    this.processorListener = processorListener;
+    this.jobCoordinator = jobCoordinator;
+    this.jobCoordinatorListener = createJobCoordinatorListener();
+    this.jobCoordinator.setListener(jobCoordinatorListener);
+  }
+
+  private StreamProcessor(String processorId, Config config, Map<String, MetricsReporter> customMetricsReporters,
+                          Object taskFactory, StreamProcessorLifecycleListener processorListener) {
+    this.taskFactory = taskFactory;
+    this.config = config;
+    this.taskShutdownMs = new TaskConfigJava(config).getShutdownMs();
+    this.customMetricsReporter = customMetricsReporters;
+    this.processorListener = processorListener;
+    this.jobCoordinator = getJobCoordinator(processorId);
+    this.jobCoordinator.setListener(createJobCoordinatorListener());
   }
 
   /**
-   * StreamProcessor Lifecycle: start()
-   * <ul>
-   * <li>Starts the JobCoordinator and fetches the JobModel</li>
-   * <li>jobCoordinator.start returns after starting the container using ContainerModel </li>
-   * </ul>
-   * When start() returns, it only guarantees that the container is initialized and submitted by the controller to
-   * execute
+   * Starts the JobCoordinator in StreamProcessor.
    */
   public void start() {
     jobCoordinator.start();
-    lifecycleListener.onStart();
   }
 
   /**
-   * Method that allows the user to wait for a specified amount of time for the container to initialize and start
-   * processing messages
+   * Stops the Streamprocessor's running components.
    *
-   * @param timeoutMs Maximum time to wait, in milliseconds
-   * @return {@code true}, if the container started within the specified wait time and {@code false} if the waiting time
-   * elapsed
-   * @throws InterruptedException if the current thread is interrupted while waiting for container to start-up
-   */
-  public boolean awaitStart(long timeoutMs) throws InterruptedException {
-    return jobCoordinator.awaitStart(timeoutMs);
-  }
-
-  /**
-   * StreamProcessor Lifecycle: stop()
-   * <ul>
-   * <li>Stops the SamzaContainer execution</li>
-   * <li>Stops the JobCoordinator</li>
-   * </ul>
+   * Can be invoked by the user of StreamProcessor api or by any of the components in StreamProcessor (JobCoordinator
+   * or SamzaContainer)
    */
-  public void stop() {
-    jobCoordinator.stop();
+  public synchronized void stop() {
+    if (container != null) {
+      LOGGER.info("Shutting down container " + container.toString() + " from StreamProcessor");
+      container.shutdown(false);
```

Comment:
Container shutdown should automatically trigger onContainerStop callback that will trigger the JobCoordinator to stop. There are multiple ways in which the StreamProcessor stops:
1. caller of StreamProcessor invokes stop()
2. Container completes processing (eg. bounded input) and shuts down
3. Container fails
4. Coordinator fails 

When either container or coordinator stops (cleanly or due to exception), it will try to shutdown the streamprocessor. This needs to be synchronized so that only one code path gets triggered for shutdown.

If container is running,
* container is shutdown cleanly and onContainerStop will trigger jobcoordinator.stop.
* container fails to shutdown cleanly -> onContainerFailure will trigger jobcoordinator.stop

If container is not running, then we should simply shutdown the jobcoordinator.